xapian-applications/omega/index_file.cc

```diff
@@ -545,6 +545,77 @@ index_add_document(const string & urlterm, time_t last_altered,
     }
 }
 
+bool
+index_update_entry(const string & urlterm,
+		   DirectoryIterator & d,
+		   Xapian::docid & did)
+{
+    if (!did) {
+	Xapian::PostingIterator p = db.postlist_begin(urlterm);
+	if (p != db.postlist_end(urlterm))
+	    did = *p;
+    }
+
+    if (did) {
+	time_t mtime = d.get_mtime();
+	time_t ctime = d.get_ctime();
+	Xapian::Document doc = db.get_document(did);
+	string value_ctime = doc.get_value(VALUE_CTIME);
+	time_t doc_ctime = binary_string_to_int(value_ctime);
+	string value_mtime = doc.get_value(VALUE_LASTMOD);
+	time_t doc_mtime = binary_string_to_int(value_mtime);
+
+	if (doc_mtime == mtime && doc_ctime <= ctime) {
+	    if (verbose)
+		cout << "Updating entry without re-extracting text" << endl;
+	    // Remove terms from the document
+	    Xapian::TermIterator term_iterator = doc.termlist_begin();
+	    while (term_iterator != doc.termlist_end()) {
+		const string& term = *term_iterator;
+		bool rem_term = startswith(term, "I") ||
+				startswith(term, "O");
+		// I: boolean filter term for "can see" permission
+		// O: Owner
+		if (rem_term) {
+		    doc.remove_term(term);
+		}
+		term_iterator++;
+	    }
```

Comment:
Again this looks correct, but there's scope to optimise as we can avoid having to iterate all the terms.  skip_to("I") will advance us to the first term >= "I", then we can `skip_to("O")` once we've handled all the `I` terms, then we can stop once we've handled all the `O` terms.

Reply:
Okay, I will try to use it